free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py

```diff
@@ -0,0 +1,290 @@
+                    nanoseconds=zt.header.stamp.nanosec
+                )
+                t = TransformStamped()
+                t.header.stamp = time.to_msg()
+                t.header.stamp
+                t.header.frame_id = namespacify(zt.header.frame_id,
+                                                self.name)
+                t.child_frame_id = namespacify(zt.child_frame_id,
+                                               self.name)
+                t.transform.translation.x = zt.transform.translation.x
+                t.transform.translation.y = zt.transform.translation.y
+                t.transform.translation.z = zt.transform.translation.z
+                t.transform.rotation.x = zt.transform.rotation.x
+                t.transform.rotation.y = zt.transform.rotation.y
+                t.transform.rotation.z = zt.transform.rotation.z
+                t.transform.rotation.w = zt.transform.rotation.w
+                self.tf_buffer.set_transform(t, f'{self.name}_RobotAdapter')
+
+        self.tf_sub = self.zenoh_session.declare_subscriber(
+            namespacify('tf', self.name),
+            _tf_callback
+        )
+
+        def _battery_state_callback(sample: zenoh.Sample):
+            battery_state = SensorMsgs_BatteryState.deserialize(sample.payload)
+            self.battery_soc = battery_state.percentage
+
+        self.battery_state_sub = self.zenoh_session.declare_subscriber(
+            namespacify('battery_state', name),
+            _battery_state_callback
+        )
+
+    def _make_random_goal_id(self):
+        return np.random.randint(0, 255, size=(16)).astype('uint8').tolist()
+
+    def _is_navigation_done(self) -> bool:
+        if self.nav_goal_id is None:
+            return True
+
+        req = NavigateToPose_GetResult_Request(goal_id=self.nav_goal_id)
+        # TODO(ac): parameterize the service call timeout
+        replies = self.zenoh_session.get(
+            namespacify('navigate_to_pose/_action/get_result', self.name),
+            zenoh.Queue(),
+            value=req.serialize(),
+            # timeout=0.5
+        )
+        for reply in replies.receiver:
+            try:
+                # Deserialize the response
+                rep = NavigateToPose_GetResult_Response.deserialize(
+                    reply.ok.payload
+                )
+                self.node.get_logger().debug(f'Result: {rep.status}')
+                if rep.status == GoalStatus.STATUS_EXECUTING:
+                    return False
+                elif rep.status == GoalStatus.STATUS_SUCCEEDED:
+                    self.node.get_logger().info(
+                        f'Navigation goal {self.nav_goal_id} reached'
+                    )
+                    return True
+                else:
+                    self.node.get_logger().error(
+                        f'Navigation goal {self.nav_goal_id} status '
+                        f'{rep.status}')
+                    return True
```

Comment:
In cases where the navigation goal has failed and the robot did not reach the goal waypoint successfully, `_is_navigation_done()` returns True. This might cause RMF to misunderstand that the robot is ready for the subsequent waypoint. It might help to either add a retry mechanism for the same goal waypoint, or trigger a replan.